### PR TITLE
add access and refresh token

### PR DESCRIPTION
**Pull request by andrey11m**

### CODE REVIEW

File: src/main/java/internship/mbicycle/storify/controller/StorifyUserController.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    public String activateEmail(@PathVariable String code, HttpServletResponse response) {
+        StorifyUser storifyUser = userService.activateEmail(code);
+        if (storifyUser.getActivationCode() == null) {
+            tokenService.setTokensAfterActivation(storifyUser);
+            response.setHeader("access_token", storifyUser.getToken().getAccessToken());
+            response.setHeader("refresh_token", storifyUser.getToken().getRefreshToken());
+            return "Thank you for registration!";
+        } else {
+            return "Activate was not successful";
+        }
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/controller/StorifyUserController.java:44]: Move logic to email-service

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @Column(name = "activation_code")
+    private String activationCode;
+
+    @Column(name = "role")
+    private String role;
+
+    @Column(name = "name")
+    private String name;
+
+    @Column(name = "tempConfirmCode")
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/model/StorifyUser.java:40]: Bad column naming

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    @Column(name = "password")
+    private String password;
+
+    @Column(name = "activation_code")
+    private String activationCode;
+
+    @Column(name = "role")
+    private String role;
+
+    @Column(name = "name")
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/model/StorifyUser.java:37]: Do you need to duplicate column name if it equals to default?

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    @Column(name = "email")
+    private String email;
+
+    @Column(name = "password")
+    private String password;
+
+    @Column(name = "activation_code")
+    private String activationCode;
+
+    @Column(name = "role")
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/model/StorifyUser.java:34]: Do you need to duplicate column name if it equals to default?

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "email")
+    private String email;
+
+    @Column(name = "password")
+    private String password;
+
+    @Column(name = "activation_code")
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/model/StorifyUser.java:31]: Do you need to duplicate column name if it equals to default?

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+public class StorifyUser implements UserDetails {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "email")
+    private String email;
+
+    @Column(name = "password")
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/model/StorifyUser.java:28]: Do you need to duplicate column name if it equals to default?

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class StorifyUser implements UserDetails {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "email")
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/model/StorifyUser.java:25]: Do you need to duplicate column name if it equals to default?

File: src/main/java/internship/mbicycle/storify/service/StorifyUserService.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+public interface StorifyUserService {
+
+    StorifyUser saveUser(StorifyUser storifyUser);
+
+    StorifyUser getUserById(long id);
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/StorifyUserService.java:13]: Use optional

File: src/main/java/internship/mbicycle/storify/service/StorifyUserService.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+    List<StorifyUser> getAllUsers();
+
+    StorifyUser activateEmail(String code);
+
+    void updateStorifyUser(StorifyUser storifyUser);
+
+    StorifyUser getUserByEmail(String email);
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/StorifyUserService.java:21]: Use optional

File: src/main/java/internship/mbicycle/storify/service/StorifyUserService.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+@Service
+public interface StorifyUserService {
+
+    StorifyUser saveUser(StorifyUser storifyUser);
+
+    StorifyUser getUserById(long id);
+
+    List<StorifyUser> getAllUsers();
+
+    StorifyUser activateEmail(String code);
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/StorifyUserService.java:17]: Bad method naming

File: src/main/java/internship/mbicycle/storify/service/TokenService.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+public interface TokenService {
+
+    String createAccessToken(StorifyUser storifyUser);
+
+    String createRefreshToken(StorifyUser storifyUser);
+
+    UsernamePasswordAuthenticationToken getAuthenticationToken(String authorizationHeader);
+
+    StorifyUser getTokensFromRefreshToken(String authorizationHeader);
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/TokenService.java:17]: Bad method naming

File: src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        userRepository.save(storifyUser);
+    }
+
+    @Override
+    public StorifyUser saveUser(StorifyUser storifyUser) {
+        storifyUser.setRole("ROLE_GUEST");
+        storifyUser.setPassword(passwordEncoder.encode(storifyUser.getPassword()));
+        storifyUser.setActivationCode(UUID.randomUUID().toString());
+        storifyUser.setToken(new Token());
+        String message = String.format(
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java:37]: Move all messages to util class

File: src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                "Hello, %s! \n " +
+                        "Welcome to Storify. Please, visit next link: http://localhost:8080/activate/%s",
+                storifyUser.getName(), storifyUser.getActivationCode()
+        );
+        mailSenderService.send(storifyUser.getEmail(), "Activation code", message);
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java:42]: Move all messages to util class

File: src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        return userRepository.save(storifyUser);
+    }
+
+    @Override
+    public StorifyUser getUserById(long id) {
+        return userRepository.findById(id).orElseThrow(() -> new UsernameNotFoundException("User not found in db"));
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java:48]: Move all messages to util class

File: src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public List<StorifyUser> getAllUsers() {
+        return userRepository.findAll();
+    }
+
+    @Override
+    public StorifyUser activateEmail(String code) {
+        StorifyUser storifyUser = userRepository.findByActivationCode(code)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found in database with it activation code"));
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java:59]: Move all messages to util class

File: src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        storifyUser.setActivationCode(null);
+        storifyUser.setRole("ROLE_USER");
+        return userRepository.save(storifyUser);
+    }
+
+    @Override
+    public StorifyUser getUserByEmail(String email) {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found in db"));
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/StorifyUserServiceImpl.java:68]: Move all messages to util class
And add identification to user. Which user not found? Petya? Misha? Vova?

File: src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+
+    private final SecurityProperties securityProperties;
+    private final StorifyUserService userService;
+
+    @Override
+    public String createAccessToken(StorifyUser storifyUser) {
+        StorifyUser userFromDb = userService.getUserByEmail(storifyUser.getEmail());
+        return JWT.create()
+                .withSubject(userFromDb.getEmail())
+                .withExpiresAt(new Date(System.currentTimeMillis() + 1000 * 60 * 5)) // 5 minutes
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java:36]: Move to util class

File: src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+                .withClaim("role", userFromDb.getRole())
+                .sign(getEncryptionAlgorithm());
+    }
+
+    @Override
+    public String createRefreshToken(StorifyUser storifyUser) {
+        return JWT.create()
+                .withSubject(storifyUser.getEmail())
+                .withExpiresAt(new Date(System.currentTimeMillis() + 1000 * 60 * 60 * 24 * 10)) // 10 days
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java:45]: Move to util class

File: src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+    private StorifyUser getStorifyUserFromAuthorizationHeader(String authorizationHeader) {
+        String token = parseTokenFromAuthorizationHeader(authorizationHeader);
+        JWTVerifier verifier = JWT.require(getEncryptionAlgorithm()).build();
+        DecodedJWT decodedJWT = verifier.verify(token);
+        String email = decodedJWT.getSubject();
+        StorifyUser storifyUser = userService.getUserByEmail(email);
+        if (securityProperties.isCheckAccess()) {
+            if (!storifyUser.getToken().getAccessToken().equals(token)) {
+                throw new RuntimeException("Tokens aren't the same");
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java:80]: Move to util class

File: src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+            }
+        }
+        return storifyUser;
+    }
+
+    private String parseTokenFromAuthorizationHeader(String authorizationHeader) {
+        if (!authorizationHeader.startsWith("Bearer ")) {
+            throw new RuntimeException("The token is incorrect");
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/TokenServiceImpl.java:88]: Move to util class

File: src/main/java/internship/mbicycle/storify/service/impl/UserDetailsServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+@RequiredArgsConstructor
+@Transactional
+public class UserDetailsServiceImpl implements UserDetailsService {
+
+    private final StorifyUserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        StorifyUser storifyUser = userRepository.findByEmail(email)
+                .orElseThrow(() -> new RuntimeException("User not found"));
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/service/impl/UserDetailsServiceImpl.java:23]: Move to util class

File: src/main/java/internship/mbicycle/storify/configuration/security/CustomAuthorizationFilter.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+
+@Slf4j
+@RequiredArgsConstructor
+public class CustomAuthorizationFilter extends OncePerRequestFilter {
+
+    private final TokenService tokenService;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        if (request.getServletPath().equals("/login") ||
```
Comment (Estrin-MBicycle) [inline src/main/java/internship/mbicycle/storify/configuration/security/CustomAuthorizationFilter.java:27]: This method looks terrible :)
Pls refactor it

**Review by Estrin-MBicycle — CHANGES_REQUESTED** [review]

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+
+import javax.persistence.*;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+
+@Entity
```
[the remark this answers is not in the document]
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:15]: Done

File: src/main/java/internship/mbicycle/storify/model/Token.java [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+@Entity
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/Token.java:15]: Done

File: src/main/java/internship/mbicycle/storify/controller/AuthenticationController.java [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+@RestController()
+@RequestMapping("/token")
+@RequiredArgsConstructor
+public class AuthenticationController {
+
+    private final TokenService tokenService;
+
+    @GetMapping("/refresh")
+    public void getNewAccessToken(HttpServletRequest request, HttpServletResponse response) {
```
[the remark this answers is not in the document]
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/controller/AuthenticationController.java:24]: Done

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java (hunk 2 above)
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:40]: Done

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java (hunk 3 above)
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:37]: Done

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java (hunk 4 above)
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:34]: Done

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java (hunk 5 above)
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:31]: Done

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java (hunk 6 above)
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:28]: Done

File: src/main/java/internship/mbicycle/storify/model/StorifyUser.java (hunk 7 above)
Comment (andrey11m) [reply src/main/java/internship/mbicycle/storify/model/StorifyUser.java:25]: Done

**Pull request merged**